examples/ucp_client_server.c

```diff
@@ -95,6 +96,95 @@ static struct {
  */
 static void usage(void);
 
+void buffer_free(ucp_dt_iov_t *iov)
+{
+    size_t idx;
+
+    for (idx = 0; idx < iov_cnt; idx++) {
+        if (iov[idx].buffer == NULL) {
+            continue;
+        }
+        mem_type_free(iov[idx].buffer);
+        iov[idx].buffer = NULL;
+        iov[idx].length = 0;
+    }
+    mem_type_free(iov);
+}
+
+ucp_dt_iov_t *buffer_malloc()
+{
+    size_t       idx;
+    ucp_dt_iov_t *iov;
+
+    iov = calloc(iov_cnt, sizeof(ucp_dt_iov_t));
+    CHKERR_ACTION(iov == NULL, "allocate memory\n", return NULL;);
+
+    for (idx = 0; idx < iov_cnt; idx++) {
+        iov[idx].length = test_string_length;
+        iov[idx].buffer = mem_type_malloc(iov[idx].length);
+        if (iov[idx].buffer == NULL) {
+            buffer_free(iov);
+            return NULL;
+        }
+        mem_type_memset(iov[idx].buffer, 0, iov[idx].length);
+    }
+
+    return iov;
+}
+
+int fill_buffer(ucp_dt_iov_t *iov)
+{
+    size_t idx;
+    int    ret  = 0;
+
+    for (idx = 0; idx < iov_cnt; idx++) {
+        ret = generate_test_string(iov[idx].buffer, iov[idx].length);
+        if (ret != 0) {
+            break;
+        }
+    }
+    CHKERR_ACTION(ret != 0, "generate test string", return -1;);
+    return 0;
+}
+
+char **copy_buffer(ucp_dt_iov_t *iov)
```

Comment:
why need to copy? looks like the created msg_str is released immediately

Reply:
@yosefe 
It keeps the same behaviour as original code.
If there's no need to keep same behavior as the original code, I could remove the related code here.
https://github.com/openucx/ucx/blob/v1.11.0-rc4/examples/ucp_client_server.c#L286-L292
```
        msg_str = calloc(1, test_string_length + 1);
        if (msg_str == NULL) {
            fprintf(stderr, "memory allocation failed\n");
            return -1;
        }

        mem_type_memcpy(msg_str, msg, test_string_length);
```

Reply:
@yosefe Do you thinks we do not need to keep the same behavior as original code?

Reply:
@yosefe If the IOV memory is from GPU/CUDA, could we use printf to print it directly?

Reply:
i see. so maybe to simplify it, only the printing function will copy to temp local buffer?

Reply:
@yosefe Got it. I'll do the corresponding change.